### ISSUE TITLE
LoRA training with modules_to_save hangs during checkpoint saving under FSDP full_shard

### ISSUE THREAD

**Issue by LagunaKi** [modelscope/ms-swift#9313]
### Checklist / 检查清单

- [x] I have searched existing issues, and this is a new bug report. / 我已经搜索过现有的 issues，确认这是一个新的 bug report。

### Problem

  When training Qwen2.5-Omni with LoRA plus `modules_to_save`, checkpoint saving hangs under FSDP.

  The training step itself can finish, but the process gets stuck during checkpoint saving. Depending on the module used in `modules_to_save`, the
checkpoint directory is either only partially written or only contains `README.md`.

  This happens even with a minimal debug run:

  - `max_steps=1`
  - `save_steps=1`
  - `per_device_train_batch_size=1`
  - `gradient_accumulation_steps=1`
  - 2 training samples
  - 2 GPUs

  
  ### Reproduction Matrix

  I ran a minimal probe with max_steps=1 and save_steps=1.

  | Case | LoRA target | modules_to_save | FSDP | Result |
  |---|---|---|---|---|
  | LLM LoRA only | LLM self-attn | none | full_shard auto_wrap | PASS |
  | Encoder LoRA only | audio/visual encoder LoRA | none | full_shard auto_wrap | PASS |
  | LLM LoRA + audio projector | LLM self-attn | thinker.audio_tower.proj | full_shard auto_wrap | HANG/TIMEOUT |
  | LLM LoRA + visual merger | LLM self-attn | thinker.visual.merger | full_shard auto_wrap | HANG/TIMEOUT |
  | LLM LoRA + both modules | LLM self-attn | both above | full_shard without auto_wrap | HANG/TIMEOUT |

  ### Observed Checkpoint Outputs

  For modules_to_save=thinker.audio_tower.proj, the checkpoint directory contains:

  README.md
  adapter_config.json
  adapter_model.safetensors
  additional_config.json
  args.json
  chat_template.jinja
  generation_config.json
  preprocessor_config.json
  processor_config.json
  spk_dict.pt
  tokenizer.json
  tokenizer_config.json
  training_args.bin

  But it does not contain:

  pytorch_model_fsdp.bin
  optimizer.bin
  scheduler.pt
  trainer_state.json
  rng_state_*.pth

  This suggests that PEFT adapter saving can complete, but FSDP model/optimizer checkpoint saving hangs afterwards.

  For modules_to_save=thinker.visual.merger, the checkpoint directory only contains:

  README.md

  This suggests the process hangs earlier, likely during PEFT adapter state dict collection or adapter saving.

  For the successful LoRA-only cases, the checkpoint directory contains the complete expected files:

  adapter_model.safetensors
  pytorch_model_fsdp.bin
  optimizer.bin
  scheduler.pt
  trainer_state.json
  rng_state_*.pth

  ### Expected Behavior

  Either:

  1. Swift should support train_type=lora + modules_to_save + FSDP full_shard and save a complete resumable checkpoint.

  Or:

  2. Swift should explicitly reject this combination during argument/config validation with a clear error message.

  ### Actual Behavior

  The training process hangs during checkpoint saving. No Python exception is raised. The job has to be killed by timeout or Slurm.

  ### Suspected Cause

  This looks related to the interaction between:

  - PEFT ModulesToSaveWrapper
  - FSDP full_shard
  - Swift/Transformers checkpoint saving
  - Accelerate FSDP model/optimizer state dict saving

  modules_to_save introduces full-rank trainable modules inside a PEFT adapter wrapper. Under FSDP, these parameters need to participate both in
  PEFT adapter state extraction and in FSDP resumable checkpoint state extraction. This path appears unstable or unsupported.

  ### Request

  Could you clarify whether modules_to_save is officially supported together with FSDP full_shard in Swift LoRA training?

  If it is supported, could you help fix the checkpoint saving hang?

  If it is not supported, it would be helpful for Swift to reject this configuration before training starts.

### Environment

  - ms-swift version: 4.0.4
  - PEFT version: 0.18.1
  - Transformers version: 5.3.0
  - Model: Qwen2.5-Omni-7B
  - Training type: `lora`
  - Tuner backend: `peft`
  - Distributed: `FSDP full_shard` and `FSDP full_shard auto_wrap`
  - Precision: `bf16`
  - GPUs: reproduced on 2 GPUs and 8 GPUs

### Configuration Pattern

  The model is trained with regular LoRA target modules, plus full-rank modules saved through `modules_to_save`.

  Example LoRA target:

  ```text
  model.layers.(\d+).self_attn.(q|k|v|o)_proj

  Example modules_to_save:

  thinker.audio_tower.proj
  thinker.visual.merger

  The Swift CLI receives equivalent arguments:

  --train_type lora
  --lora_rank 64
  --lora_alpha 128
  --lora_dropout 0.05
  --lora_dtype bfloat16
  --target_regex 'model.layers.(\d+).self_attn.(q|k|v|o)_proj'
  --modules_to_save thinker.audio_tower.proj thinker.visual.merger
  --aligner_lr 1e-5
  --fsdp '{"fsdp": "full_shard auto_wrap", "fsdp_config": {"activation_checkpointing": true, "backward_prefetch": "backward_pre",
"forward_prefetch": false, "transformer_layer_cls_to_wrap": ["Qwen2_5OmniDecoderLayer", "Qwen2_5OmniVisionBlock", "Qwen2_5OmniAudioEncoderLayer"],
"use_orig_params": true, "cpu_ram_efficient_loading": true}}'



### Additional Information / 补充信息

_No response_
